kore/src/main/scala/org/kframework/definition/transformers.scala

```diff
@@ -116,4 +115,24 @@ class DefinitionTransformer(moduleTransformer: Module => Module) extends (Defini
   }
 }
 
+object KViz {
+  def from(f: java.util.function.UnaryOperator[K], name:String):KViz = KViz(f(_), name)
+
+  def apply(f: K => K, name: String): KViz = f match {
+    case f: KViz => f
+    case _ => new KViz(f, name)
+  }
+}
 
+class KViz(f: K => K, name: String) extends (K => K) {
+  override def apply(input: K): K = {
+    input match {
+      case c: KToken  => f(c)
+      case tc: KApply =>
+        f(tc)
+        tc.items.forEach(apply)
+        tc
+      case _ => throw new AssertionError("Not expected downed term in visitor " + name + " term: " + input)
```

Comment:
what's a downed term?

Reply:
Good question.
After parsing, we transform the parser classes into KApps with references to productions.
We then apply `down` to replace `KApply(KLabel("#KRewrite"), L, R)` into `KRewrite(L, R)`.
[Here](https://github.com/runtimeverification/k/blob/master/kore/src/main/scala/org/kframework/kore/interface.scala#L31) are the classes. We do this because it's easier to write transformers in kompile. But for the LSP I need all my terms to have references to the original production.
So instead of 
`Cache.save(down(trm))` and `Cache.load()` I had to modify it to
`Cache.save(trm)` and `down(Cache.load())`.